addons/mail/static/src/models/message_reaction_per_emoji/message_reaction_per_emoji.js

```diff
@@ -0,0 +1,79 @@
+/** @odoo-module **/
+
+import { registerNewModel } from '@mail/model/model_core';
+import { attr, many2one, one2many} from '@mail/model/model_field';
+import { clear, insert, insertAndReplace, link, replace, unlink, unlinkAll } from '@mail/model/model_field_command';
+
+function factory(dependencies) {
+
+    class MessageReactionPerEmoji extends dependencies['mail.model'] {
+
+        //----------------------------------------------------------------------
+        // Public
+        //----------------------------------------------------------------------
+
+        /**
+         *  Removes the reaction for current user by clicking the message emoji reaction.
+         */
+        removeReaction() {
+            this.message.onEmojiSelection(this.emoji_unicode);
+        }
+
+        //----------------------------------------------------------------------
+        // Private
+        //----------------------------------------------------------------------
+
+        /**
+         * @override
+         */
+        static _createRecordLocalId(data) {
+            return `${this.modelName}_${data.messageId}_${data.emoji_unicode}`;
+        }
+
+        /**
+         * @private
+         */
+        _computeHasUserReacted() {
+            const reactionForCurrentUser = this.messaging.models['mail.message_reaction'].all().filter(r =>
+                    r.message.id === this.messageId && r.emoji_unicode === this.emoji_unicode && r.reactor === this.messaging.currentPartner)[0];
+            if (reactionForCurrentUser) {
+                return true;
+            }
+            return false;
+        }
+
+        /**
+         * @private
+         */
+        _computeMessage() {
+            return insert({ id: this.messageId });
+        }
+    }
+
+    MessageReactionPerEmoji.fields = {
+        count: attr({
+            required: true,
+        }),
+        emoji_unicode: attr({
+            required: true,
+        }),
+        hasUserReacted: attr({
+            compute: '_computeHasUserReacted',
+            default: false,
+        }),
+        message: many2one('mail.message', {
+            compute: '_computeMessage',
+            inverse: 'messageReactionsPerEmoji',
+            required: true,
+        }),
+        messageId: attr({
+            required: true,
+        }),
+    };
+
+    MessageReactionPerEmoji.modelName = 'mail.message_reaction_per_emoji';
```

Comment:
The concept of a "message reaction" is good, in that it maps 1 reaction to 1 message / 1 part.

Here the intent of this model is make a group of reactions on a message of all users that use the same type of reaction. I would use the term "emoji", because reaction with "emoji" is just the sole mean to do reaction but it shouldn't necessarily be limited to that, and the concept is broader than emoji.

I suggest keeping the name simple with something like `MessageReactionGroup` (or `MessageReactionSet` or `MessageReactionCollection`, or... I think you get it, just define a "cohort"). This naming is less intimidating, and better show that the modelling is about grouping reactions according to some criteria.

It's okay to simplify modelling by assuming grouping is always per same reaction, same reaction = having same content.